src/adapter/src/coord.rs

```diff
@@ -1635,12 +1578,191 @@ impl Coordinator {
     }
 
+    /// Invokes the optimizer on all indexes and materialized views in the catalog and inserts the
+    /// resulting dataflow plans into the catalog state.
+    ///
+    /// `ordered_catalog_entries` must by sorted in dependency order, with dependencies ordered
+    /// before their dependants.
+    ///
+    /// This method does not perform timestamp selection for the dataflows, nor does it create them
+    /// in the compute controller. Both of these steps happen later during bootstrapping.
+    fn bootstrap_dataflow_plans(
+        &mut self,
+        ordered_catalog_entries: &[CatalogEntry],
+    ) -> Result<(), AdapterError> {
+        // The optimizer expects to be able to query its `ComputeInstanceSnapshot` for
+        // collections the current dataflow can depend on. But since we don't yet install anything
+        // on compute instances, the snapshot information is incomplete. We fix that by manually
+        // updating `ComputeInstanceSnapshot` objects to ensure they contain collections previously
+        // optimized.
+        let mut instance_snapshots = BTreeMap::new();
+
+        let optimizer_config = OptimizerConfig::from(self.catalog().system_config());
+
+        for entry in ordered_catalog_entries {
+            let id = entry.id();
+            match entry.item() {
+                CatalogItem::Index(idx) => {
+                    // Collect optimizer parameters.
+                    let compute_instance =
+                        instance_snapshots.entry(idx.cluster_id).or_insert_with(|| {
+                            self.instance_snapshot(idx.cluster_id)
+                                .expect("compute instance exists")
+                        });
+
+                    // The index may already be installed on the compute instance. For example,
+                    // this is the case for introspection indexes.
+                    if compute_instance.contains_collection(&id) {
+                        continue;
+                    }
+
+                    // Build an optimizer for this INDEX.
+                    let mut optimizer = optimize::index::Optimizer::new(
+                        self.owned_catalog(),
+                        compute_instance.clone(),
+                        entry.id(),
+                        optimizer_config.clone(),
+                    );
+
+                    // MIR ⇒ MIR optimization (global)
+                    let index_plan = optimize::index::Index::new(entry.name(), &idx.on, &idx.keys);
+                    let global_mir_plan = optimizer.optimize(index_plan)?;
+                    let optimized_plan = global_mir_plan.df_desc().clone();
+
+                    // MIR ⇒ LIR lowering and LIR ⇒ LIR optimization (global)
+                    let global_lir_plan = optimizer.optimize(global_mir_plan)?;
+                    let physical_plan = global_lir_plan.df_desc().clone();
+                    let metainfo = global_lir_plan.df_meta().clone();
```

Comment:
We could avoid these `clone`s if `GlobalLirPlan<Unresolved>` supported `unapply`. I understand that it doesn't do that because we want to make sure that timestamp selection is always ran as part of the optimization, but:

* As this commit shows, it doesn't actually ensure that.
* We need `DataflowDescription` and metainfo here, so we can store them in `CatalogPlans`. We could also return `(GlobalMirPlan, GlobalLirPlan)` from `bootstrap_dataflow_plans` and populate `CatalogPlans` later, but that just seems like unnecessary hassle?

Reply:
It's not that easy, because there are two different `GlobalLirPlan` types, for indexes and for MVs. So we'd have to return `(BTreeMap<GlobalId, index::GlobalLirPlan>, BTreeMap<GlobalId, materialized_view::GlobalLirPlan>)` (or probably some helper struct that wraps these two). That might not be too bad, but then we would still end up calling `set_optimized_plan` at a different place than `set_physical_plan`, making it harder to keep track of where plans are set. Feels a bit icky to me, but if you feel strongly I'll change it like you suggested.

> I think chances are that whoever does this will look for callers of `set_as_of` and will append their code right after these calls

I mean, people can try to make all kinds of bad changes, but I'd hope could review would prevent this from getting merged!

Reply:
If we do that I suggest to completely remove the `Unresolved` / `Resolved` distinction in `GlobalLirPlan` and the `resolve` method from `optimize/materialized_view.rs` / `optimize/index.rs` and just say that "timestamp selection is not part of the optimizer pipeline" for these two statement types (in other words the pipeline can be completed without timestamp selection).

However somebody needs to remember that when this statement changes also the above code should be reverted.

Reply:
This seems fine to me too. If we want to start making index/MV optimizations based on timestamp selection, we will make it part of the optimizer pipeline again, which would break `bootstrap_dataflow_plans` because there wouldn't be a way to get the physical plan without supplying a timestamp. So at this point we'd be forced to reconsider out approach here (probably to just optimize up to the MIR plan and leaving the rest for later).